src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeGraph.cs

```diff
@@ -205,137 +203,336 @@ public RuntimeGroup(ITaskItem item)
+                        if (ApplyVersionsToParent)
                         {
-                            var versionedRid = runtimeDescription.RuntimeIdentifier;
+                            imports.Add(CreateRuntime(Parent, version: version, qualifier: qualifier));
+                        }
 
-                            var archIndex = versionedRid.IndexOf('-');
+                        yield return new RIDMapping(CreateRuntime(BaseRID, version: version, qualifier: qualifier), imports);
 
-                            if (archIndex != -1)
+                        foreach (var architecture in Architectures)
+                        {
+                            // base + version + architecture + qualifier =>
+                            //      base + version + qualifier, 
+                            //      base + version + architecture, 
+                            //      base + version
+                            //      base + lastVersion + architecture + qualifier
+                            //      parent + version + architecture + qualifier (optionally)
+                            var archImports = new List<RID>()
                             {
-                                versionedRid = versionedRid.Substring(0, archIndex);
-                            }
+                                CreateRuntime(BaseRID, version: version, qualifier: qualifier),
+                                CreateRuntime(BaseRID, version: version, architecture: architecture),
+                                CreateRuntime(BaseRID, version: version),
+                                CreateRuntime(BaseRID, version: lastVersion, architecture: architecture, qualifier: qualifier)
+                            };
 
-                            int insertAt = 0;
-
-
-                            while (insertAt < imports.Count && imports[insertAt].StartsWith(versionedRid))
+                            if (ApplyVersionsToParent)
                             {
-                                insertAt++;
+                                imports.Add(CreateRuntime(Parent, version: version, architecture: architecture, qualifier: qualifier));
                             }
 
-                            imports.Insert(insertAt, runtimeDescription.RuntimeIdentifier);
+                            yield return new RIDMapping(CreateRuntime(BaseRID, version: version, architecture:architecture, qualifier: qualifier), archImports);
                         }
-                        else
+
+                        if (TreatVersionsAsCompatible)
                         {
-                            imports.Add(runtimeDescription.RuntimeIdentifier);
+                            lastVersion = version;
                         }
+                    }
+                }
+            }
+
+            private bool IsNullOrRoot(string rid)
+            {
+                return rid == null || rid == rootRID;
```

Comment:
I'm a bit confused at the "rootRID".  In the code above, it is "any".  But in the [runtime.json file](https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.NETCore.Platforms/runtime.json), it seems that "any" imports "base".

I would think "base" is the rootRID.

Reply:
`any` is the true root.  `base` was introduced as a private RID that only we would use back when we tried to use runtime.json to lift up our package versions to a minimum version https://github.com/dotnet/buildtools/blob/96ee49b1f47f1a61dc040b2be2d976bbaa7903c4/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs#L195-L207.  This technique never really worked perfectly so we stopped using it, then NuGet broke it (intentionally) by allowing downgrade of package versions referenced by runtime.json if the version referenced was higher in the graph than the package which brought in the runtime package.